### ISSUE TITLE
Support Azure Trusted Signing certificate on NuGet.org account settings

### ISSUE THREAD

**Issue by jozefizso** [NuGet/NuGetGallery#10027]
### NuGet Product(s) Involved

Other/NA

### The Elevator Pitch

The Azure Trusted Signing service rotates the managed authenticode certificates each three days.

As the NuGet Gallery requires to specify a single, long lived certificate, it is not feasible to sign nuget packages with the ATS.

### Additional Context and Details

@dlemstra has authored a proposal design to fix this problem: https://github.com/NuGet/Home/pull/13791

**Comment by joelverhagen** [comment]
Thanks for opening the issue @jozefizso. I agree using Trusted Signing with NuGet.org is a painful process today. You need to extract the .cer file from your signed artifact (or get it from Trusted Signing directly, perhaps) and then manually upload it to your account profile on NuGet.org. 

For my own experiments I created a CLI tool to extract the .cer file and include it as a GitHub Actions artifact.
Tool: https://www.nuget.org/packages/Knapcode.CertificateExtractor
Workflow: https://github.com/joelverhagen/PackageLifeCycle/blob/7930b64f7e72518faf6b646187b47b46e2d64d22/.github/workflows/build.yml#L76-L87

This is better than nothing, but not much better!

**Comment by dlemstra** [comment]
I would nice if we could get a new option to specify the `Enhanced key usage` that is shown on the page of our Certificate Profile inside our Trusyed Signing Page. That is the same one as the bottom one that is shown here:

![Enhanced key usage](https://github.com/NuGet/NuGetGallery/assets/10426229/85e04f50-3437-4ffd-96c9-7db37cee19c6)

And then when we upload the NuGet package the server checks if this matches and also checks if the signing of the certificate is done by the right authority.

**Comment by ianjmcm** [comment]
Supporting the Subscriber identity validation EKU as @dlemstra points out would be durable over all certificate rotations/renewals. Here is the public docs on these values: https://learn.microsoft.com/en-us/azure/trusted-signing/concept-trusted-signing-cert-management#subscriber-identity-validation-eku.

**Comment by mjcheetham** [comment]
This problem is currently impacting Git Credential Manager that ships a NuGet package that's signed by Azure Trusted Signing – would love to see some support for these sorts of certificates/signed packages, and a solution that is durable over all of the short-lived certs.

**Comment by jmecosta** [comment]
> Thanks for opening the issue @jozefizso. I agree using Trusted Signing with NuGet.org is a painful process today. You need to extract the .cer file from your signed artifact (or get it from Trusted Signing directly, perhaps) and then manually upload it to your account profile on NuGet.org.
> 
> For my own experiments I created a CLI tool to extract the .cer file and include it as a GitHub Actions artifact. Tool: https://www.nuget.org/packages/Knapcode.CertificateExtractor Workflow: https://github.com/joelverhagen/PackageLifeCycle/blob/7930b64f7e72518faf6b646187b47b46e2d64d22/.github/workflows/build.yml#L76-L87
> 
> This is better than nothing, but not much better!

@joelverhagen is the & "bin\Sign.Cli\sign.exe" something that you have created? can you share it? or do you have any pointers how to get nuget sign with trusted service? or are you using https://github.com/dotnet/sign/pull/716

**Comment by dlemstra** [comment]
The `"bin\Sign.Cli\sign.exe"` executable is the output of the `dotnet/sign` project. The branch of my PR is still work in progress so it's subject to change. I have been using it locally but not yet in a pipeline.

**Comment by jmecosta** [comment]
yep, just getting to that conclusion also, there are a few cli options changed but pretty much what im looking for... thank you

**Comment by dlemstra** [comment]
And those options might still change because the PR is still being reviewed by the team.

**Comment by jmecosta** [comment]
Yep important is the sign functionality itself if that works we can start integrating it already

**Comment by joelverhagen** [comment]
@dlemstra is spot on. I have a private build of sign CLI used in the pipeline links above.

**But to be clear this GitHub issue tracks better support of Trusted Signing on NuGet.org** (_verification_ of signing at push time in NuGet.org) and not the _sign_ flow which is tracked by https://github.com/dotnet/sign/issues/683.

**Comment by glennawatson** [comment]
Has anyone seen issues where the validation fails with sign cli and trusted signing. 

I'm extracting out the certificate using @joelverhagen tool

I can see the valid certificate in the nuget package explorer tool

seeing

```
The package was signed, but the signing certificate (SHA-1 thumbprint <thumbprint>) is not associated with your account. You must register this certificate to publish signed packages.
```

I have got the SHA1/256 hash and confirmed its there, and in the cert area it's showing microsoft signing there like its picked up and processed the cert.

I have tried the sign cli experimental version, jsign client, signtool as recommended in the docs and all producing the same results.

**Comment by jmecosta** [comment]
no problems so far, you could try this
dotnet tool install --global AzureTrustedSignTool
dotnet tool install Knapcode.CertificateExtractor --global
dotnet tool install --global sign --prerelease
AzureTrustedSignTool sign --folder ./pkgs --searchpattern *.nupkg --accountname YourAccount --profilename YourSignProfile

that signs the packages and generates a .cer next to each nuget. so you can upload both at same time. So far no problems with this approach.

**Comment by glennawatson** [comment]
@jmecosta I think its some sort of special rules for my package rather than a process issue. 

Eg having some packages get published (with the same single owner) and some not. 

Splat had the DotNetFoundation as a owner and I know there are special rules for DNF owned packages, also its in a reserved space.

I suspect it's more rules around that rather than a legitimate certificate failure given the intermitient results about which ones pass (ones never owned by the DotNetFoundation user owner)

**Comment by jmecosta** [comment]
Interesting I've never notice certificate bounded to package names. 
 If so the error message is not really accurate

**Comment by jmecosta** [comment]
@glennawatson i can confirm this also, ive just experience the same. the account is shared, and ive uploaded the CER file with my account and the other user was not able to publish the nuget... 

he then re-upload the CER to his account and was  able to do so. so conclusion is that the ones uploading the nugets need to have the CER files uploaded also

**Comment by joelverhagen** [comment]
@jmecosta, @glennawatson, @mjcheetham - folks, thank you for your input so far in this issue.

@dlemstra has opened a proposal design to address this NuGet.org limitation. Any input you have for Dirk and our team would be most appreciated over on the proposal PR: https://github.com/NuGet/Home/pull/13791!

**Comment by ErikEJ** [comment]
Would love to see this happen, as an open source developer all other alternatives involve manually signing on your local PC, which is not substainable with NuGet.org and CI, as code signing on Nuget.org is all or nothing.

**Comment by jozefizso** [comment]
As open source developers we really need this feature.

**Comment by virzak** [comment]
In my experience, this is a minor annoyance - at most a few minutes of extra work per day. I run this script on the days I need to generate my [sync-method-generator](https://github.com/zompinc/sync-method-generator). It creates a single certificate which is then uploaded to nuget manually (not sure if there's a way to upload certificate using APIs).


```ps1
$env:AZURE_TENANT_ID = "redacted"
$env:AZURE_CLIENT_SECRET = "redacted"
$env:AZURE_CLIENT_ID = "redacted"

$version = dotnet dnx --yes nbgv get-version --format json | jq -r '.NuGetPackageVersion'

Write-Output "Version: $version"

Remove-Item "$PSScriptRoot\packages" -Recurse -Force -Confirm:$false
dotnet pack -o packages
dotnet dnx -y --prerelease sign code trusted-signing --base-directory "$PSScriptRoot/packages" "Zomp.SyncMethodGenerator.$version.nupkg" --trusted-signing-endpoint "https://eus.codesigning.azure.net" --trusted-signing-account "zomp" --trusted-signing-certificate-profile "zomp-main" -v normal

dotnet dnx --prerelease --yes Knapcode.CertificateExtractor --file "$PSScriptRoot/packages/Zomp.SyncMethodGenerator.$version.nupkg" --output packages --all

# The following was generated by ChatGPT
# Get all .cer files in current directory (you can change the path)
$certFiles = Get-ChildItem -Path "$PSScriptRoot\packages" -Filter *.cer

# Find the one with subject containing "Zomp"
$zompCert = $certFiles | Where-Object {
  try {
    $cert = New-Object System.Security.Cryptography.X509Certificates.X509Certificate2($_.FullName)
    $cert.Subject -like '*Zomp*'
  }
  catch {
    $false
  }
}

if ($zompCert) {
  # Keep this certificate and delete all others
  foreach ($file in $certFiles) {
    if ($file.FullName -ne $zompCert.FullName) {
      Remove-Item $file.FullName -Force
      Write-Host "Deleted: $($file.Name)"
    }
    else {
      Write-Host "Kept: $($file.Name)"
    }
  }
}
else {
  Write-Host "No certificate found with subject containing 'Zomp'"
}
```

**Comment by fflaten** [comment]
> In my experience, this is a minor annoyance - at most a few minutes of extra work per day

Great that you found a workaround that suits your manual release flow.

For continous deployment pipelines, any manual step isn't just an annoyance but a showstopper.

**Comment by jmecosta** [comment]
also showstoper if there are multiple sub orgs, and those sub orgs dont have access to upload the certficates (they can get apis to upload nugets)

**Comment by mmsmits** [comment]
Is there any update on this?

I agree a manual release approach might work here, but I really need this to be done automatically using a CI/CD pipeline. 

If there's a Nuget API call to register a new certificate to my organization, I can get around it, but I can't upload a new certificate every three days.

**Comment by dlemstra** [comment]
I have written a short story about the workaround that I used: https://dlemstra.github.io/github-stories/2026/magick-net-now-uses-nuget-trusted-publishing-together-with-artifact-signing/

p.s. Trusted Signing has been rebranded as Artifact Signing and is now available for everyone.

**Comment by jozefizso** [comment]
> I have written a short story about the workaround that I used: [dlemstra.github.io/github-stories/2026/magick-net-now-uses-nuget-trusted-publishing-together-with-artifact-signing](https://dlemstra.github.io/github-stories/2026/magick-net-now-uses-nuget-trusted-publishing-together-with-artifact-signing/)
> 
> p.s. Trusted Signing has been rebranded as Artifact Signing and is now available for everyone.

I love the Python script to extract the certificate information. Although the [Knapcode.CertificateExtractor](https://www.nuget.org/packages/Knapcode.CertificateExtractor) is a great tool, it's quite a big dependency.

**Comment by ErikEJ** [comment]
@dlemstra I cannot see that it is available for Individual Developers outside the US !?

**Comment by martincostello** [comment]
If it is, [the documentation](https://learn.microsoft.com/en-us/azure/artifact-signing/faq#what-if-my-country-region-is-not-listed-in-country-region-drop-down-on-the-identity-validation-page) hasn't been updated either:

> For Public Trust certificates, Artifact Signing is currently available to organizations in the USA, Canada, the European Union, and the United Kingdom, **as well as individual developers in the USA and Canada**.

**Comment by ErikEJ** [comment]
@martincostello Yeah, just tried to add a new Individual account and only US regions were available.  😢

**Comment by mmsmits** [comment]
Still no update on this? Even after the official release?

**Comment by jozefizso** [comment]
Please, can you implement this? This is such a pain release nuget packages which **MUST** be signed, because Nuget **REQUIRES IT** based on my profile where I have some previous certificates set.

**Comment by jmecosta** [comment]
It’s really demotivating having this ticket open for so long. We are being told we are in the 'age of AI' and being pushed to automate every part of our development lifecycle—yet, for something as fundamental as automated package publishing, we are still stuck with this. It’s hard to reconcile the 'AI-first' corporate messaging with the reality that a feature so critical to modern automated publishing has been buried in a backlog for 2 years.

**Comment by jozefizso** [comment]
Can we get an official response here? We are releasing our packages in the most secure pay, paying for Azure Artifacts Siging $10/month and we **MUST** manually upload certificates to the NuGet Gallery.

**Comment by glennawatson** [comment]
The trend in the industry is to go for increasingly shorter timelines on certificates. There's talk of 47 day max lifetime. 

Unless we get some sort of ability to do oidc upload or similar ability it'll be just more than a inconvenience on the entire platform .

https://www.digicert.com/blog/understanding-the-new-code-signing-certificate-validity-change

**Comment by jmecosta** [comment]
@glennawatson those azure trusted certs validaty are 3 days, they rotate them daily if not mistaken ---- in that sense the problem lies in publishing the cert for upload since we dont publish nugets every day, which brakes automated flows.

i think the industry will face quite a big push back if they reduce the validaty of those certs to 47 days, since that would break patch upgrades totally for all customers... Well basically it kind of all alligns to the premise that all gravitates towards Microsoft store black hole. So wouldnt be surprised

**Comment by jozefizso** [comment]
Missing support for Azure Trusted Singing certificates in NuGet Gallery creates huge friction for us as we cannot automate our build and release processes.

**Comment by Frulfump** [comment]
The design PR is merged so at least that is done.
https://github.com/NuGet/Home/pull/13791

**Comment by jozefizso** [comment]
> The design PR is merged so at least that is done. [NuGet/Home#13791](https://github.com/NuGet/Home/pull/13791)

on Mar 31, 2025

**Comment by cyotek** [comment]
> API keys are strongly discouraged for automated publishing and should be replaced with [Trusted Publishing](https://aka.ms/nuget/trusted-publishing), which is currently supported for GitHub Actions, with broader CI/CD support coming soon. We strongly recommend updating now to Trusted Publishing to ensure a seamless transition. For publishing from command line or unsupported CI/CD workflows, [API keys](https://www.nuget.org/account/apikeys?forceApiKeys=true) continue to work.

So you're "discouraging" the use of API Keys on NuGet in favour of this but after two years you still haven't truly added support for it? Maybe _you_ want everyone to use GitHub Actions but certainly my aspect of everyone has no desire to.

**Comment by glennawatson** [comment]
> > API keys are strongly discouraged for automated publishing and should be replaced with [Trusted Publishing](https://aka.ms/nuget/trusted-publishing), which is currently supported for GitHub Actions, with broader CI/CD support coming soon. We strongly recommend updating now to Trusted Publishing to ensure a seamless transition. For publishing from command line or unsupported CI/CD workflows, [API keys](https://www.nuget.org/account/apikeys?forceApiKeys=true) continue to work.
> 
> So you're "discouraging" the use of API Keys on NuGet in favour of this but after two years you still haven't truly added support for it? Maybe _you_ want everyone to use GitHub Actions but certainly my aspect of everyone has no desire to.

Trusted publishing works. https://learn.microsoft.com/en-us/nuget/nuget-org/trusted-publishing

That's not what this issue is about

**Comment by cyotek** [comment]
> > > API keys are strongly discouraged for automated publishing and should be replaced with [Trusted Publishing](https://aka.ms/nuget/trusted-publishing), which is currently supported for GitHub Actions, with broader CI/CD support coming soon. We strongly recommend updating now to Trusted Publishing to ensure a seamless transition. For publishing from command line or unsupported CI/CD workflows, [API keys](https://www.nuget.org/account/apikeys?forceApiKeys=true) continue to work.
> > 
> > 
> > So you're "discouraging" the use of API Keys on NuGet in favour of this but after two years you still haven't truly added support for it? Maybe _you_ want everyone to use GitHub Actions but certainly my aspect of everyone has no desire to.
> 
> Trusted publishing works. https://learn.microsoft.com/en-us/nuget/nuget-org/trusted-publishing
> 
> That's not what this issue is about

Hmm, apologies. I assumed they were related - I logged into NuGet to do a package update and found my key had expired. But in order to generate it I had to jump hoops past "Use Trusted ~~Signing~~ Publishing", and _then_ upload the certificate extracted from the package. To me it's all part of the same (frustrating workflow).

I now realise I stopped reading after "Trusted" and just jumped to conclusions - sorry!

**Comment by jozefizso** [comment]
Even the Trusted Publishing is neglected. It is supported for GitHub Actions only.

Using GitLab, Azure DevOps, or other services mis still unsupported.

**Comment by jmecosta** [comment]
@jozefizso maybe that should be tracked in seperate request then - we dont want microsfot to spend another 2 years implementing this one refers to :)

**Comment by dansiegel** [comment]
There's been a few industry scoped comments here... but I'm going to be real for a moment. The NuGet team started several years ago on this mission to get Package authors like myself to buy into the notion that I should sign the package before uploading to NuGet.org. Back then it was a little bit of initial effort but not really a big deal since most bigger package authors were also MVPs getting free certs from DigiCert at the time. Now that I have to pay for my cert I have avoided publishing any package updates because the costs are insane, well over $1,000 to sign NuGet packages that I continue to give away for free. This is a bad deal. The Cert industry is beyond greedy, SSL.com wants $500 just to verify that you're using Azure Key Vault for your cert, and that's not including the cost of the cert itself.

If the NuGet team is serious about providing a rich ecosystem full of trusted, signed packages, this issue needs to move from "Next Quarter" to P1 fix and out next week.

Azure Artifact Signing isn't perfect by any stretch, but the $9.99 price point is at least reasonable until the NuGet team can come up with something better that could integrate with the Trusted Package Publishing in the same way Google Play takes the signing effort away from developers now.

**Comment by virzak** [comment]
Follow-up to my earlier comment: the signing tool can export the certificate directly, so the extractor and the filtering block go away.

```ps1
dotnet dnx SignUniversal.Cli@1.0.34 --yes sign "$PSScriptRoot/packages/Zomp.SyncMethodGenerator.$version.nupkg" `
  --export-certificate "$PSScriptRoot/packages/signing.cer" `
  --trusted-signing-endpoint "https://eus.codesigning.azure.net" `
  --trusted-signing-account "zomp" --trusted-signing-certificate-profile "zomp-main"
```

It comes from the process that just signed, so no subject matching is needed; its SHA-256 matches the fingerprint `dotnet nuget verify` reports, which is what the gallery matches on. Still a manual upload each rotation, so the ask here stands.

**Comment by jmecosta** [comment]
<img width="1199" height="502" alt="Image" src="https://github.com/user-attachments/assets/d253e3ab-d929-4738-85ce-9c5b04ddb2df" />


I guess this will not get implemented. as it will not be needed

**Comment by martincostello** [comment]
Trusted signing is unrelated to trusted publishing

**Comment by jmecosta** [comment]
i know, but if you use trusted signing in github, and trusted publishing in github. then you dont need this implementation for anything. Since you should be able to publish nuget packages directly from there - im assuming, that road does not require you to register the short live certifcates everytime you sign a package (but i havent tried) in nuget.org.

**Comment by martincostello** [comment]
> im assuming, that road does not require you to register the short live certifcates everytime you sign a package (but i havent tried) in nuget.org

I far as I'm aware, that is still the case - this issue is tracking the request to remove that need

**Comment by jmecosta** [comment]
ah ok, then nevermind - sucks

**Comment by Jaxelr** [comment]
Worth mentioning that trusted signing is not even called trusted signing anymore, so maybe a title edit might make this more discoverable moving forward?

**Comment by jmecosta** [comment]
Maybe they will change the name again before implementing this

**Comment by freddydk** [comment]
Looks like this is due today - does anybody know whether this is being worked on?

**Comment by jmecosta** [comment]
the baby is late

**Comment by Jaxelr** [comment]
> Maybe they will change the name again before implementing this

You betcha!

**Comment by mmsmits** [comment]
@joelverhagen Would it speed things up of I try to add a PR to fix this?

**Comment by ErikEJ** [comment]
@mmsmits give it a try!

**Comment by mmsmits** [comment]
> [@mmsmits](https://github.com/mmsmits) give it a try!

Will do!

**Comment by mmsmits** [comment]
@joelverhagen @ErikEJ 

I want to agree on the approach with you before I open a PR.

**Where the logic would go**

- The thumbprint check happens in the ProcessSignature job (`SignatureValidator.ValidatePackageRegistrationSigningRequirementsAsync` → `PackageRegistrationExtensions.IsAcceptableSigningCertificate`). The gallery itself never parses uploaded `.cer` files. So the Artifact Signing detection and linking of the Durable Identity Value (DIV) would live in the job. That is also where `SignaturePartsExtractor` already writes certificate details back to the gallery DB.
- `IsSigningAllowed` / `IsSigningRequired` / `IsAcceptableSigningCertificate` would also count linked DIVs. Without this, removing a user's certificate link (as the design describes) would switch `IsSigningRequired` to false and quietly weaken the signing policy.
- As discussed in the design review, the new behavior would be behind a flight, so each step can ship on its own for selected accounts.

**Proposed schema (new EF6 migration)**

- `DurableIdentityValues`: `Key`, `Value` (EKU OID string, unique), `Subject`, `ShortSubject`, `Issuer`, `ShortIssuer`, `LastSeen`.
- `UserDurableIdentityValues`: join table on `UserKey` + `DurableIdentityValueKey`, the same pattern as `UserCertificates`. One DIV can be linked to several users/orgs, as agreed in the design review.

**Linking rules (from the design)**

A DIV is linked only when all of these are true:
- the author-signed package's end certificate matches a certificate registered to an owner (or to the required signer),
- the certificate chains to `Microsoft ID Verified Code Signing PCA 2021` (with no assumption about how many intermediate CAs are in between), issued by `Microsoft Identity Verification Root Certificate Authority 2020`,
- the certificate has the Artifact Signing EKU `1.3.6.1.4.1.311.97.1.0` and a DIV EKU (`1.3.6.1.4.1.311.97.*`),
- the package signature's timestamp is within the configured window (default 30 days, set in the job's appsettings).

The trusted PCA/root thumbprints and the OIDs would also be configuration values. Link, unlink and add actions would get audit records and telemetry, matching how certificates work today.

**Proposed PR split**

1. Schema + extension-method changes + linking and matching in the ProcessSignature job (design step 1), behind the flight, with unit tests.
2. Account page: a separate "Artifact Signing identities" table with a remove button (step 2).
3. A form to add a DIV by hand, copied from the Azure portal (step 3).

**Open questions**

1. Naming: the design review chose "Durable Identity Value" so the concept stays independent of Artifact Signing. Do you want the tables named after that (`DurableIdentityValues`, as above), or after the service (`ArtifactSigningIdentities`)?
2. When a DIV is linked, should the matching certificate be unlinked from the account (as the design says), or kept until step 2 ships so users still see something in their account?
3. Which flight should gate this: a new flight in the gallery's feature-flag system that the job reads, or a job-only flag with an account allow list?

Thanks! I'll wait for your go-ahead before I open anything.

**Comment by ErikEJ** [comment]
@dlemstra thoughts on the implementation proposal?
